precompile/precompiles.go

```diff
@@ -1,24 +1,114 @@
+	govprecompile "github.com/cosmos/evm/precompiles/gov"
+	"github.com/cosmos/evm/precompiles/p256"
+	slashingprecompile "github.com/cosmos/evm/precompiles/slashing"
+	stakingprecompile "github.com/cosmos/evm/precompiles/staking"
+	evmkeeper "github.com/cosmos/evm/x/vm/keeper"
+
+	channelkeeper "github.com/cosmos/ibc-go/v10/modules/core/04-channel/keeper"
+
 	pauth "github.com/xpladev/xpla/precompile/auth"
 	pbank "github.com/xpladev/xpla/precompile/bank"
-	pdistribution "github.com/xpladev/xpla/precompile/distribution"
-	pstaking "github.com/xpladev/xpla/precompile/staking"
 	pwasm "github.com/xpladev/xpla/precompile/wasm"
 )
 
+const bech32PrecompileBaseGas = 6_000
+
 var PrecompiledAddressesXpla = []common.Address{
-	pbank.Address, pstaking.Address, pdistribution.Address, pwasm.Address, pauth.Address,
+	pbank.Address, pwasm.Address, pauth.Address,
 }
 
-func RegistPrecompiledContract(ak pwasm.AccountKeeper, bk pbank.BankKeeper, sk pstaking.StakingKeeper, dk pdistribution.DistributionKeeper, wms pwasm.WasmMsgServer, wk pwasm.WasmKeeper, authAk pauth.AccountKeeper) {
-	vm.PrecompiledContractsPrague[pbank.Address] = pbank.NewPrecompiledBank(bk)
-	vm.PrecompiledContractsPrague[pstaking.Address] = pstaking.NewPrecompiledStaking(sk)
-	vm.PrecompiledContractsPrague[pdistribution.Address] = pdistribution.NewPrecompiledDistribution(dk)
-	vm.PrecompiledContractsPrague[pwasm.Address] = pwasm.NewPrecompiledWasm(ak, wms, wk)
-	vm.PrecompiledContractsPrague[pauth.Address] = pauth.NewPrecompiledAuth(authAk)
+// NewAvailableStaticPrecompiles returns the list of all available static precompiled contracts from Cosmos EVM.
+// NOTE: this should only be used during initialization of the Keeper.
+func NewAvailableStaticPrecompiles(
+	stakingKeeper stakingkeeper.Keeper,
+	distributionKeeper distributionkeeper.Keeper,
+	channelKeeper *channelkeeper.Keeper,
+	evmKeeper *evmkeeper.Keeper,
+	govKeeper govkeeper.Keeper,
+	slashingKeeper slashingkeeper.Keeper,
+	evidenceKeeper evidencekeeper.Keeper,
+	ak pwasm.AccountKeeper,
+	bk pbank.BankKeeper,
+	wms pwasm.WasmMsgServer,
+	wk pwasm.WasmKeeper,
+	authAk pauth.AccountKeeper,
+	codec codec.Codec,
+) map[common.Address]vm.PrecompiledContract {
+	// Clone the mapping from the latest EVM fork.
+	precompiles := maps.Clone(vm.PrecompiledContractsPrague)
+
+	// secp256r1 precompile as per EIP-7212
+	p256Precompile := &p256.Precompile{}
+
+	bech32Precompile, err := bech32.NewPrecompile(bech32PrecompileBaseGas)
+	if err != nil {
+		panic(fmt.Errorf("failed to instantiate bech32 precompile: %w", err))
+	}
+
+	stakingPrecompile, err := stakingprecompile.NewPrecompile(stakingKeeper)
+	if err != nil {
+		panic(fmt.Errorf("failed to instantiate staking precompile: %w", err))
+	}
+
+	distributionPrecompile, err := distprecompile.NewPrecompile(
+		distributionKeeper,
+		stakingKeeper,
+		evmKeeper,
+	)
+	if err != nil {
+		panic(fmt.Errorf("failed to instantiate distribution precompile: %w", err))
+	}
+
```

Comment:
Is there a reason why ics20precompile was removed?

Reply:
Because it has a dependency of evm/erc20 module currently, so I think I need to dig deep into more.

Reply:
The ics20 precompile has a dependency of ibc transfer keeper, and ibc transfer keeper DOES have a dependency of erc20 keeper, so..